arlo_server/__init__.py

```diff
@@ -1218,36 +1221,82 @@ def audit_report(election_id):
             ["Round {:d} End".format(round.round_num), round.ended_at]
         )
 
-        ballots = (
+        ballot_draws = (
             SampledBallotDraw.query.filter_by(round_id=round.id)
             .join(SampledBallotDraw.batch)
-            .add_entity(Batch)
             .filter_by(jurisdiction_id=jurisdiction.id)
             .order_by("batch_id", "ballot_position")
             .all()
         )
+        all_sampled_ballot_draws += ballot_draws
 
         report_writer.writerow(
             [
                 "Round {:d} Samples".format(round.round_num),
                 " ".join(
                     [
                         "(Batch {:s}, #{:d}, Ticket #{:s})".format(
-                            batch.name, b.ballot_position, b.ticket_number
+                            b.batch.name, b.ballot_position, b.ticket_number
                         )
-                        for b, batch in ballots
+                        for b in ballot_draws
                     ]
                 ),
             ]
         )
 
+    if election.online:
+        report_writer.writerow(["All Sampled Ballots"])
+        report_writer.writerow(
+            ["Ballot", "Ticket Numbers", "Audited?", "Audit Result", "Comments"]
+        )
+        # Write a row for each ballot that looks like this:
+        # "Batch 1, #13",Round 1: 0.123,Audited,some_candidate_id,A comment
+        # The Ticket Numbers column is a bit tricky:
+        # If a ballot was sampled multiple times in a round: Round 1: 0.123, 0.456
+        # If a ballot was sampled in multiple rounds: Round 1: 0.123, Round 2: 0.456
+
+        # First group all the ballot draws by the actual ballot
+        for _, ballot_draws in group_by(
```

Comment:
I can see how using the `group_by` here in code vs the DB might make it easier to generate the Tickets fields the way you want, but then I would consider joining, in the database, with `SampledBallot`, otherwise you'll be doing an extra query every time you reference `b.sampled_ballot`.

Reply:
This is very helpful! I was wondering how all the foreign objects I needed were magically available haha. Guess I could have thought a little harder about that.

I also didn't think super hard about doing the grouping in the database, since I was just reusing the data that was already being queried. Do you think it's worth thinking through a better query that does some aggregation? Part of me thinks it's not worthwhile since we need the unaggregated ballot samples as well for the existing report behavior.

Reply:
Added join here: e334e08